File: 2015/erlang/day_03.erl
-module(day_03).
-export([main/0]).

main() ->
    {ok, Binary} = file:read_file("../input/day-03.txt"),
    Chars = string:trim(binary_to_list(Binary)),

    Start = {0, {0, 0}, {0, 0}, {0, 0}, [{0, 0}], [{0, 0}]},
    {_Turn, _Single, _Santa, _Robo, Part1Houses, Part2Houses} =
        lists:foldl(fun step/2, Start, Chars),

    io:fwrite("Part 1: ~p~n", [length(lists:uniq(Part1Houses))]),
    io:fwrite("Part 2: ~p~n", [length(lists:uniq(Part2Houses))]).

step(Ch, {Turn, Single, Santa, Robo, P1Houses, P2Houses}) ->
    NewSingle = move(Single, Ch),
    case Turn rem 2 =:= 0 of
        true ->
            NewSanta = move(Santa, Ch),
            {Turn + 1, NewSingle, NewSanta, Robo,
             [NewSingle | P1Houses], [NewSanta | P2Houses]};
        false ->
            NewRobo = move(Robo, Ch),
            {Turn + 1, NewSingle, Santa, NewRobo,
             [NewSingle | P1Houses], [NewRobo | P2Houses]}
    end.

move({X, Y}, $>) -> {X + 1, Y};
move({X, Y}, $<) -> {X - 1, Y};
move({X, Y}, $^) -> {X, Y + 1};
move({X, Y}, $v) -> {X, Y - 1}.

File: 2015/perl/day-03.pl
use 5.42.0;
open my $fh, '<', '../input/day-03.txt';

sub advance( $houses, $char ) {
    my %delta = (
        '>' => { x =>  1, y =>  0 },
        '<' => { x => -1, y =>  0 },
        'v' => { x =>  0, y =>  1 },
        '^' => { x =>  0, y => -1 },
    );

    my %next = @$houses[-1]->%*;
    %next = (
        x => $next{x} + $delta{$char}{x},
        y => $next{y} + $delta{$char}{y} );

    push @$houses, \%next;
}

sub count_houses($houses) {
    my %hash;
    for my $house (@$houses) {
        my $key = sprintf '%d|%d', $house->{x}, $house->{y};
        $hash{$key}++;
    }
    scalar keys %hash;
}

my $origin = { x => 0, y => 0 };
my @houses = ($origin);
my @santa  = ($origin);
my @robo   = ($origin);
my $turn   = 0;

while ( defined( my $char = getc($fh) ) ) {
    next unless $char =~ /[<>v^]/;
    advance \@houses, $char;
    advance \@santa,  $char if $turn % 2 == 0;
    advance \@robo,   $char if $turn % 2 == 1;
    $turn++;
}

say 'Part 1: ', count_houses( \@houses );
say 'Part 2: ', count_houses( [ @santa, @robo ] );
